Deploy Automation: Refactoring abort code

markbiek opened the pull request:
### Changes proposed in this Pull Request:

We had a lot of duplicate code surrounding error conditions where we switched back to the original branch and aborted execution.

This moves that code into a common function that we use everywhere.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Related to #1096.

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested. Please, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

To ensure nothing breaks, we need to follow the testing instructions for the following PRs:

- #1122
- #1109
- #1115

<!-- End testing instructions -->

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

markbiek committed:
Refactor abort code

We moved the code to display a message, switch back to a branch, and
abort into a single function we can use everywhere.

markbiek committed:
Better handle when pushing the tag fails

daledupreez commented on scripts/commands/readme.js:
```diff
@@ -18,12 +17,11 @@ async function updateReadMe( currentBranchName ) {
-			'The file readme.txt does not exist. Verify you are in the wc-calypso-bridge project directory.'
+		return abortAndSwitchToBranch(
+			'The file readme.txt does not exist. Verify you are in the wc-calypso-bridge project directory.',
+			'error',
```
Might it be worth using constants for the `error`|`warning`|`info` strings?
```suggestion
			NOTICE_LEVEL_ERROR,
```

daledupreez replied:
Might it be worth using constants for the `error`|`warning`|`info` strings?
```suggestion
			NOTICE_LEVEL_ERROR,
```

daledupreez reviewed (commented):
I have some comments up from inspection.

daledupreez reviewed (commented):
There is a bug/TODO in `scripts/commands/readme.js`: the final line has the following content:
```js
updateReadMe( '1098-update/readme-section' );
```
We shouldn't be hard-coding the branch name.

daledupreez commented on scripts/commands/build.js:
```diff
@@ -33,18 +33,12 @@ async function buildRelease( currentBranchName ) {
 	if ( isDevBuild() ) {
-		error(
-			"You may have a dev build. Please make sure you aren't running 'npm start'"
+		return abortAndSwitchToBranch(
+			"You may have a dev build. Please make sure you aren't running 'npm start'",
+			'error',
+			currentBranchName
 		);
```
I kept tripping over this error due to my `build` directory existing from some older build process - it may be worth sharing a command to work around this, even if it's just `rm -r build`.

daledupreez commented on scripts/commands/build.js:
```diff
@@ -62,17 +56,18 @@ async function buildRelease( currentBranchName ) {
 	info( 'Creating a new release build. This make take some time...' );
 	const interval = setInterval( () => process.stdout.write( '.' ), 1000 );
 
 	await execAsync( 'npm i' );
```
Should we run `npm ci` here to make sure we're installing the full set of dependencies and don't have any unexpected additions/deviations? See the [npm-ci docs](https://docs.npmjs.com/cli/v8/commands/npm-ci) for more details.

daledupreez commented on scripts/commands/build.js:
```diff
@@ -86,17 +81,24 @@ async function buildRelease( currentBranchName ) {
+		);
 	}
 
 	// Delete the node_modules directory so they don't get packaged up.
```
It would be helpful if we tagged on a message to note that users need to (re)run `npm install` after we've deleted `node_modules`.

daledupreez reviewed (commented):
Testing overall went well, though I did find some small areas where the code or messages could be better in general.

markbiek committed:
Switch to running npm ci on build

This is better suited for automated scripts

markbiek committed:
Improved error message when a dev build is detected

markbiek committed:
Improved how we build up the verification error string

markbiek committed:
Info message letting people know that we've deleted node_modules/

markbiek committed:
Switch to using constants for notice levels

markbiek committed:
fixup! Improved how we build up the verification error string

markbiek committed:
Removed debug code for readme command

This command will be worked into the main flow once we merge the version bump command

daledupreez commented on scripts/utils.js:
```diff
@@ -41,6 +48,22 @@ export async function promptContinue( msg ) {
 }
 
+export async function abortAndSwitchToBranch( msg, type, branchName = null ) {
+	if ( type === 'error' ) {
+		error( msg );
+	} else if ( type === 'warning' ) {
+		warning( msg );
+	} else {
+		info( msg );
+	}
```
Minor nit: we could also use `NOTICE_LEVEL` in this code.
```suggestion
	if ( type === NOTICE_LEVEL.INFO ) {
		error( msg );
	} else if ( type === NOTICE_LEVEL.WARNING ) {
		warning( msg );
	} else {
		info( msg );
	}
```

daledupreez reviewed (approved):
I've inspected the changes from yesterday, and I think this is ready to merge, as the changes weren't major.

daledupreez commented:
> There is a bug/TODO in `scripts/commands/readme.js`: the final line has the following content:
> 
> ```js
> updateReadMe( '1098-update/readme-section' );
> ```
> 
> We shouldn't be hard-coding the branch name.

@markbiek, do we have a tracking issue for fixing/addressing this?

markbiek commented:
> > There is a bug/TODO in `scripts/commands/readme.js`: the final line has the following content:
> > ```js
> > updateReadMe( '1098-update/readme-section' );
> > ```
> >    
> > We shouldn't be hard-coding the branch name.
> 
> @markbiek, do we have a tracking issue for fixing/addressing this?

@daledupreez I'm assuming you mean the work we need to do to tie together the version bump and readme update functionality? Assuming that's the case, it's covered in #1099.

markbiek committed:
Update scripts/utils.js

Co-authored-by: daledupreez <dale@automattic.com>
